From 7234032294158956c7bd2fa6f48e72b29cba644e Mon Sep 17 00:00:00 2001
From: "Lu, Xingjiang" <xingjiang.lu@intel.com>
Date: Mon, 4 Mar 2024 18:31:12 +0800
Subject: [PATCH] Use a private thread to replace the public one

Use a private thread to replace the public one, to prevent the
reconnection failure when pm.installer want to bind the installd from
ServiceManager.

In case of pm.installer fail to initialize caused by installd service
not ready, there was a WA patch from AOSP and add a reconnection
callback which handled by BackgroundThread. But sometimes in cycling
test, this public thread might be busy and handling the tasks from other
services, then pm.installer can't reconnect installd before timeout.

Also, AudioServer/AudioPolicyService depends on installd has been
registered to ServiceManager successfully. Increase the timeout value of
TimeCheck for AudioPolicyService might avoid the fatal exception to be
occurred.

TEST: 500 cycles per S3/S5 test.


Tracked-on: OAM-116189
Signed-off-by: Xie, Hongcheng <hongcheng.xie@intel.com>
Signed-off-by: Lu, Xingjiang <xingjiang.lu@intel.com>
---
 .../core/java/com/android/server/pm/Installer.java  | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/services/core/java/com/android/server/pm/Installer.java b/services/core/java/com/android/server/pm/Installer.java
index 13c5bb0230e2..fc9d6b784e51 100644
--- a/services/core/java/com/android/server/pm/Installer.java
+++ b/services/core/java/com/android/server/pm/Installer.java
@@ -178,9 +178,16 @@ public class Installer extends SystemService {
             }
         } else {
             Slog.w(TAG, "installd not found; trying again");
-            BackgroundThread.getHandler().postDelayed(() -> {
-                connect();
-            }, CONNECT_RETRY_DELAY_MS);
+            new Thread(new Runnable() {
+                public void run() {
+                    try {
+                        Thread.sleep(CONNECT_RETRY_DELAY_MS);
+                        connect();
+                    } catch (Exception e) {
+                        Slog.e(TAG, "found exception " + e + " in reconnect()");
+                    }
+                }
+            }).start();
         }
     }

--
2.34.1

